tokio-postgres/tests/test/copy_both.rs

```diff
@@ -0,0 +1,91 @@
+use futures::{SinkExt, StreamExt, TryStreamExt};
+use std::future;
+use tokio_postgres::NoTls;
+use tokio_postgres::SimpleQueryMessage::Row;
+
+// https://stackoverflow.com/a/35907071
+fn contains(haystack: &[u8], needle: &[u8]) -> bool {
+    haystack
+        .windows(needle.len())
+        .any(|window| window == needle)
+}
+
+#[tokio::test]
+async fn copy_both() {
+    copy_both_impl(true).await;
+    copy_both_impl(false).await;
+}
+
+async fn copy_both_impl(graceful: bool) {
+    // form SQL connection
+    let conninfo = "host=127.0.0.1 port=5433 user=postgres replication=database";
+    let (client, connection) = tokio_postgres::connect(conninfo, NoTls).await.unwrap();
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("connection error: {}", e);
+        }
+    });
+
+    client
+        .simple_query("DROP TABLE IF EXISTS foo")
+        .await
+        .unwrap();
+    client
+        .simple_query("CREATE TABLE foo (i text)")
+        .await
+        .unwrap();
+
+    let slot = client
+        .simple_query("CREATE_REPLICATION_SLOT slot TEMPORARY LOGICAL \"test_decoding\"")
+        .await
+        .unwrap();
+
+    let lsn = match &slot[0] {
+        Row(row) => row.get("consistent_point").unwrap(),
+        _ => panic!("unexpeced query message"),
+    };
+
+    client
+        .simple_query("INSERT INTO foo VALUES ('fde61ce315faac78b17c')")
+        .await
+        .unwrap();
+
+    let query = format!("START_REPLICATION SLOT slot LOGICAL {}", lsn);
+    let duplex_stream = client
+        .copy_both_simple::<bytes::Bytes>(&query)
+        .await
+        .unwrap();
+
+    let (mut sink, stream) = duplex_stream.split();
+
+    // We can only do some basic sanity checking of the raw stream. We filter for messages starting
+    // with 'w', which are XLogData messages, and check that they contain the transaction we expect
+    let actual: Vec<_> = stream
+        .try_filter(|buf| future::ready(buf[0] == b'w'))
```

Comment:
Looks like this is causing a test failure.

Reply:
It seems to be working now, can you test again?

Reply:
Failure is in CI: https://github.com/sfackler/rust-postgres/pull/778/checks?check_run_id=2689467390